src/sage/matrix/matrix2.pyx

```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
         else:
             return subspace
 
+    def hermitian_decomposition(self):
+        r"""
+        Return the conjugate square root factorization of a Hermitian matrix.
+
+        Generally the base ring needs to be a finite field `GF(q**2)`.
+
+        OUTPUT:
+
+        For a matrix `A` the routine returns a matrix `B` such that,
+
+        .. MATH::
+
+            A = B^\ast B
+
+        where `B^\ast` is the conjugate-transpose.
+
+        ALGORITHM:
+
+        The algorithm is a translation of BaseChangeToCanonical in the GAP `forms` package for a Hermitian form.
+
+        TESTS:
+
+        Verify we can decompose a Hermitian matrix::
+
+            sage: A = matrix(GF(11**2),[[1,4,7],[4,1,4],[7,4,1]])
+            sage: B = A.hermitian_decomposition()
+            sage: A == B*B.H
+            True
+
+        Verify we can decompose a Hermatrix matrix for which there is no triangular decomposition::
+
+            sage: A = matrix(GF(3**2),[[0, 1, 2], [1, 0, 1], [2, 1, 0]])
+            sage: B = A.hermitian_decomposition()
+            sage: A == B*B.H
+            True
+
+        """
+        from sage.matrix.constructor import identity_matrix
+        from sage.misc.functional import sqrt
+
+        F = self._base_ring
+        n = self.nrows()
+
+        if not (F.is_finite() and F.order().is_square()):
+            raise ValueError("The base ring must be a finite field GF(q**2).")
+
+        q = sqrt(F.order())
+
+        def conj_square_root(u):
+            if u == 0:
+                return 0
+            z = F.multiplicative_generator()
+            k = u.log(z)
+            if k % (q+1) != 0:
+                raise ValueError("Unable to factor: u is not in base field GF(q)")
+            return z**(k // (q+1))
+
+        if not self.is_hermitian():
+            raise ValueError("matrix is not Hermitian")
+
+        A = copy(self)
+        D = identity_matrix(F, n)
+        row = 0
+
+        # Diagonalize A
+        while True:
+            row += 1
+
+            # Look for a non-zero element on the main diagonal, starting from `row`
+            i = row - 1  # Adjust for zero-based indexing in Sage
```

Comment:
It is probably better to make `row` 0-based.

Reply:
I think we can just get rid of i. I'll look tomorrow.